pkg/aggregator/tags/store.go

```diff
@@ -0,0 +1,181 @@
+			tags: tagsBuffer.Copy(),
+			refs: 1,
+		}
+		tc.tagsByKey[key] = entry
+		tc.cap++
+		tc.telemetry.miss.Inc()
+	}
+
+	return entry
+}
+
+// Shrink will try to release memory if cache usage drops low enough.
+func (tc *Store) Shrink() {
+	stats := entryStats{}
+	for key, entry := range tc.tagsByKey {
+		if entry.refs == 0 {
+			delete(tc.tagsByKey, key)
+		} else {
+			stats.visit(entry)
+		}
+	}
+
+	if len(tc.tagsByKey) < tc.cap/2 {
+		new := make(map[ckey.TagsKey]*Entry, len(tc.tagsByKey))
+		for k, v := range tc.tagsByKey {
+			new[k] = v
+		}
+		tc.cap = len(new)
+		tc.tagsByKey = new
+	}
+
+	tc.updateTelemetry(&stats)
+}
+
+func (tc *Store) updateTelemetry(s *entryStats) {
+	t := &tc.telemetry
+
+	tlmMaxEntries.Set(float64(tc.cap), t.name)
+	tlmEntries.Set(float64(len(tc.tagsByKey)), t.name)
+
+	for i := 0; i < 3; i++ {
+		tlmTagsetRefsCnt.Set(float64(s.refsFreq[i]), t.name, fmt.Sprintf("%d", i+1))
+	}
+	for i := 3; i < 8; i++ {
+		tlmTagsetRefsCnt.Set(float64(s.refsFreq[i]), t.name, fmt.Sprintf("%d", 1<<(i-1)))
+	}
+
+	tlmTagsetMinTags.Set(float64(s.minSize), t.name)
+	tlmTagsetMaxTags.Set(float64(s.maxSize), t.name)
+	tlmTagsetSumTags.Set(float64(s.sumSize), t.name)
+}
+
+func newCounter(name string, help string, tags ...string) telemetry.Counter {
+	return telemetry.NewCounter("aggregator_tags_store", name,
+		append([]string{"cache_instance_name"}, tags...), help)
+}
+
+func newGauge(name string, help string, tags ...string) telemetry.Gauge {
+	return telemetry.NewGauge("aggregator_tags_store", name,
+		append([]string{"cache_instance_name"}, tags...), help)
+}
+
+var (
+	tlmHits          = newCounter("hits_total", "number of times cache already contained the tags")
+	tlmMiss          = newCounter("miss_total", "number of times cache did not contain the tags")
+	tlmEntries       = newGauge("entries", "number of entries in the tags cache")
+	tlmMaxEntries    = newGauge("max_entries", "maximum number of entries since last shrink")
+	tlmTagsetMinTags = newGauge("tagset_min_tags", "minimum number of tags in a tagset")
+	tlmTagsetMaxTags = newGauge("tagset_max_tags", "maximum number of tags in a tagset")
+	tlmTagsetSumTags = newGauge("tagset_sum_tags", "total number of tags stored in all tagsets by the cache")
+	tlmTagsetRefsCnt = newGauge("tagset_refs_count", "distribution of usage count of tagsets in the cache", "ge")
+)
+
+type storeTelemetry struct {
+	hits telemetry.SimpleCounter
+	miss telemetry.SimpleCounter
+	name string
+}
+
+func newStoreTelemetry(name string) *storeTelemetry {
+	return &storeTelemetry{
+		hits: tlmHits.WithValues(name),
+		miss: tlmMiss.WithValues(name),
+		name: name,
+	}
+}
+
+type entryStats struct {
+	refsFreq [8]uint64
+	minSize  int
+	maxSize  int
+	sumSize  int
+}
+
+func (s *entryStats) visit(e *Entry) {
+	r := e.refs
+	if r < 4 {
+		s.refsFreq[r-1]++
+	} else if r < 64 {
+		s.refsFreq[bits.Len64(r)]++ // Len(4) = 3, Len(63) = 6
+	} else {
+		s.refsFreq[7]++
+	}
+
+	n := len(e.tags)
+	if n < s.minSize {
```

Comment:
As `s.minSize` is not initialized, the value is always `0` and `n < s.minSize` is always false.